### DISCUSSION TITLE
Seeking General Criticism and Feedback for Java.net HTTP Server

### DISCUSSION THREAD

**Discussion by evanr1234** [community/community#147821] [Programming Help]
### Body

Hi Guys,

Newbie here, my plan is to put this HTTP server on Amazon Web Services somehow, with the React/frontEnd on my A2hosting. I need it to be powerful enough to handle high traffic, maybe thousands of concurrent users.

Can this server handle that kind of traffic? Does anyone have any suggestions on how to optimize this? What happens if there are more users than it can handle?

Thank you,
Evan

`public class ServerRequests {
    
    public static final int http = 80;
    public static final String CRLF = "\r\n";
    ServerSocket s;
    
    
    protected ServerSocket getServerSocket(int port) throws Exception{
        
        return new ServerSocket(port);
        
        
    }
    
    
    public void runServer(int port) throws Exception{
        
        s = getServerSocket(port);
        
        while(true) {
            
            try {
                System.out.println("server is running");
                Socket us = s.accept();
                Handler(us);
                
                
            } catch (IOException e) {
                System.err.println(e);
            }
            
        }

    }
    public void Handler(Socket s) throws Exception, MessagingException {

        new Thread(() -> {
        
        try {
            
       final InputStream is;
       final PrintWriter os;
       BCryptPasswordEncoder PE = new BCryptPasswordEncoder();
      BCryptPasswordEncoder PE2 = new BCryptPasswordEncoder();
      
           MimeMessage mesg;
           Session session;
           
            String from = s.getInetAddress().toString();
        System.out.println("Accepted connection from " + from);
        is = s.getInputStream();
          os = new PrintWriter(s.getOutputStream());

        byte[] buffer = new byte[1024];
        int bytesRead;
        Pattern p;
        Pattern p2;
        Pattern p3;
        Pattern p4;
        Pattern p5;
        Pattern p6;
        Pattern p7;
        Pattern p8;
        Pattern p9;
        Pattern p10;
        Pattern p11;
        
        Matcher m;
        Matcher m2;
        Matcher m3;
        Matcher m4;
        Matcher m5;
        Matcher m6;
        Matcher m7;
        Matcher m8;
        Matcher m9;
        Matcher m10;
        Matcher m11;
        
        while ((bytesRead = is.read(buffer)) != -1) {
           String data = new String(buffer, 0, bytesRead);
         String SP1 = "[$](four).*";
         String SP2 = "[&](two).*";
         String SP3 = "[@](one).*";
         String SP4 = "[&](thr).*";
         String SP5 = "[&](reg).*";
         String SP6 = "[&](log).*";
         String SP7 = "[&](email).*";
        String SP8 = "[&](E).*";
        String SP9 = "[&](pdit).*";
        String SP10 = "[&](D).*";
        String SP11 = "[&](K).*";
        
        
            
             p = Pattern.compile(SP1);
             p2 = Pattern.compile(SP2);
             p3 = Pattern.compile(SP3);
             p4 = Pattern.compile(SP4);
             p5 = Pattern.compile(SP5);
             p6 = Pattern.compile(SP6);
             p7 = Pattern.compile(SP7);
             p8 = Pattern.compile(SP8);
             p9 = Pattern.compile(SP9);
             p10 = Pattern.compile(SP10);
             p11 = Pattern.compile(SP11);
             
             m = p.matcher(data);
             m2 = p2.matcher(data);
             m3 = p3.matcher(data);
             m4 = p4.matcher(data);
             m5 = p5.matcher(data);
             m6 = p6.matcher(data);
             m7 = p7.matcher(data);
             m8 = p8.matcher(data);
             m9 = p9.matcher(data);
             m10 = p10.matcher(data);
             m11 = p11.matcher(data);
           

                        
              
              
              
             
if (m.find()) {
    

    
int num1 = m.group(0).indexOf("{");
int num2 = m.group(0).indexOf("}");

        JsonArray jso = new getReplies().getReplies2(m.group(0).substring(num1+7, num2-1));
        IncrementViews IV = new IncrementViews();
        IV.inc(m.group(0).substring(num1+7, num2-1));
        
            System.out.println("here is new jso: " + jso + "from substring: " + 
                    m.group(0).substring(num1, num2));
                 os.print("HTTP/1.1 200 OK" + CRLF);
              os.print("Content-Type: text/plain; charset=utf-8" + CRLF);
             os.print("Access-Control-Allow-Origin:*" + CRLF);
             os.print("Access-Control-Allow-Headers: Content-Type" + CRLF);           
            os.print("Server: http://localhost:80/" + CRLF);
            
        os.print(CRLF);
        os.print(jso);
       os.flush();
       os.close();
    
       System.out.println("String receieved 12/7: " + jso);

} else if(m2.find()) {System.out.println("m2");

int in1 = m2.group(0).indexOf("{");
int in2 = m2.group(0).lastIndexOf("}");
System.out.println(m2.group(0).substring(in1+7, in2-1));

System.out.println("this is jso2: ");
JsonArray jso2 = new fetchTopics().fetchTopics2(m2.group(0).substring(in1+7, in2-1));

               os.print("HTTP/1.1 200 OK" + CRLF);
              os.print("Content-Type: text/plain; charset=utf-8" + CRLF);
             os.print("Access-Control-Allow-Origin:*" + CRLF);
             os.print("Access-Control-Allow-Headers: Content-Type" + CRLF);           
            os.print("Server: http://localhost:80/" + CRLF);
            
        os.print(CRLF);
        
        os.print(jso2);

 os.flush(); //new addition
       os.close();

} //new addition

else if(m3.find()) {System.out.println("m3");

int m3int = m3.group(0).indexOf("{");
int m3int2 = m3.group(0).indexOf("}");
int m3int4 = m3.group(0).indexOf("myData");
int m3int3 = m3.group(0).lastIndexOf("}");
int m3int5 = m3.group(0).indexOf("@", 2);
int m3int6 = m3.group(0).lastIndexOf("@");
int m3int9 = m3.group(0).indexOf("@", m3int5 + 2);
int m3int7 = m3.group(0).indexOf("-");
int m3int8 = m3.group(0).lastIndexOf("-");

String m31 = m3.group(0).substring(m3int+7, m3int2-1);
//topic
String m32 = m3.group(0).substring(m3int4+9, m3int3-1);
//reply
String m33 = m3.group(0).substring(m3int5+1, m3int9);
//date
String m34 = m3.group(0).substring(m3int7+1, m3int8);
//user
System.out.println("topic: " + m31 + "    reply " + m32 + "     date: " + m33
+ "    user " + m34);

insertReply it = new insertReply();
it.insertIt(m33, m31, m34, m32);

 //incrementLastPostT ILP = new incrementLastPostT();
    //ILP.inc(m31, m33);

updateTopicDate UT = new updateTopicDate();
UT.update(m33, m31);

incrementReplies IR = new incrementReplies();

IR.inc(m31);

 os.flush(); //new addition
       os.close();}

else if(m4.find()) {
    
 
    
    int ind2 = m4.group(0).lastIndexOf("@");
    int ind3 = m4.group(0).indexOf("@");
    int ind4 = m4.group(0).indexOf("@", ind3 + 1);
    int ind5 = m4.group(0).indexOf("{");
    int ind6 = m4.group(0).lastIndexOf("}");
    int ind7 = m4.group(0).indexOf("-");
    int ind8 = m4.group(0).lastIndexOf("-");
     int ind9 = m4.group(0).indexOf("=");
    int ind10 = m4.group(0).lastIndexOf("=");
    
    String g1 = m4.group(0).substring(5, ind4);
    String g2 = m4.group(0).substring(ind4+1, ind2);
    String g3 = m4.group(0).substring(ind5 + 7, ind6 -1);
    String g4 = m4.group(0).substring(ind7+1, ind8);
    String g5 = m4.group(0).substring(ind9+1, ind10);

System.out.println("from: " + m4.group(0) + "value one: " + g1 + " value 2:  " + g2 +
        "value 3: " + g3 + "value 4: " + g4 + "value5: " + g5);

insertTopic IT = new insertTopic();
IT.insertIt(g1, g2, g3, g4, g5);
        
    incrementTopics IT2 =  new incrementTopics();
   IT2.inc(g3);
    
    incrementLastPostT ILP = new incrementLastPostT();
    ILP.inc(g3, g4);

    System.out.println("running ILP with: " + g3 + "and " + g4);
 os.flush(); //new addition
       os.close();}

else if (m5.find()) {
    
    Properties props = new Properties();
    
     SecureRandom random = new SecureRandom();
     
   int randomInt = random.nextInt(100000000);
        
        int m5int = m5.group(0).indexOf("%");
    int m5int2 = m5.group(0).lastIndexOf("%");
    int m5int3 = m5.group(0).indexOf("=");
    int m5int4 = m5.group(0).lastIndexOf("=");
  
    int m5int7 = m5.group(0).indexOf("@");
    int m5int8 = m5.group(0).lastIndexOf("@");
    String m5s = m5.group(0).substring(m5int+1, m5int2);
    String m5s2 = m5.group(0).substring(m5int3+1, m5int4);
    
    String m5s4 = m5.group(0).substring(m5int7+1, m5int8);

   String BCm5s2 = PE.encode(m5s2);
   
    insertUser IU = new insertUser();
    
   IU.insertIt("1", m5s, BCm5s2, m5s4, from, randomInt);
    System.out.println("UN: " + m5s + "PW: " 
             + m5s2 + "email: " + m5s4 + "IP: " + from);
          Authenticator auth = new Authenticator() {
            
            @Override
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication("[un]", "[pw]"); 
            }
            
        };
    
        props.put("mail.smtp.host","mail.pharmacyhq.com");
        props.put("mail.smtp.socketFactory.port", "465");
        props.put("mail.smtp.socketFactory.port", "javax.net.ssl.SSLSocketFactory");
        props.put("mail.smtp.auth", "true");
        props.put("mail.smtp.port", "465");
        props.put("mail.smtp.starttls.enable","true"); 
        props.put("mail.smtp.ssl.enable","true"); 

    session = Session.getDefaultInstance(props, auth);
    session.setDebug(true);

        mesg = new MimeMessage(session);
        mesg.setFrom(new InternetAddress("forum@pharmacyhq.com"));
        InternetAddress toAddress = new InternetAddress(m5s4);
        mesg.addRecipient(Message.RecipientType.TO, toAddress);
        mesg.setSubject("Subject");
        mesg.setText("PharmacyHQ Discussion Forum "
                + " -"+ "please click to confirm email:  " + m5s + "  "  + "http://localhost:3000/ConfEmail2/"+ randomInt);
        
        Transport.send(mesg);
        
    System.out.println("email sent");
}

else if(m6.find()) {
 
    int m6i = m6.group(0).indexOf("=");
     int m6i2 = m6.group(0).lastIndexOf("=");
     int m6i3 = m6.group(0).indexOf("-");
     
     String user = m6.group(0).substring(m6i+1, m6i2);
     String PW = m6.group(0).substring(m6i2+1, m6i3);

     userFound uf = new userFound();
     String use = uf.found(user, PW, PE);
     
     System.out.println("PW recieved: " + PW);
     
                os.print("HTTP/1.1 200 OK" + CRLF);
              os.print("Content-Type: text/plain; charset=utf-8" + CRLF);
             os.print("Access-Control-Allow-Origin:*" + CRLF);
             os.print("Access-Control-Allow-Headers: Content-Type" + CRLF);           
            os.print("Server: http://localhost:80/" + CRLF);
            
        os.print(CRLF);
        os.print(use);
     
        os.flush(); //new addition
       os.close();
     
    
} else if(m7.find()) {
    
    
      int m7i = m7.group(0).indexOf("{");
     int m7i2 = m7.group(0).lastIndexOf("}");
     
     String emailCode = m7.group(0).substring(m7i+7, m7i2-1);

    int EC = Integer.parseInt(emailCode);
    System.out.println("emailCode String: " + emailCode);
    
    confirmEmail CE = new confirmEmail();
    CE.conf(EC);

} else if (m8.find()) {
    
    int m8i = m8.group(0).indexOf("=");
    int m8i2 = m8.group(0).lastIndexOf("=");
    
    String user1 = m8.group(0).substring(m8i+1, m8i2);
    
    returnID RID = new returnID();
    String id3 = RID.found(user1);
    
              Authenticator auth = new Authenticator() {
            
            @Override
            protected PasswordAuthentication getPasswordAuthentication() {
                
                return new PasswordAuthentication("[un], "[pw]");
                
            }
            
        };
    
                Properties props = new Properties();
        props.put("mail.smtp.host","mail.pharmacyhq.com");
        props.put("mail.smtp.socketFactory.port", "465");
        props.put("mail.smtp.socketFactory.port", "javax.net.ssl.SSLSocketFactory");
        props.put("mail.smtp.auth", "true");
        props.put("mail.smtp.port", "465");
        props.put("mail.smtp.starttls.enable","true"); 
        props.put("mail.smtp.ssl.enable","true"); 

    session = Session.getDefaultInstance(props, auth);
    session.setDebug(true);

         mesg = new MimeMessage(session);
        
        mesg.setFrom(new InternetAddress("forum@pharmacyhq.com"));
        
        InternetAddress toAddress = new InternetAddress(user1);
        
        mesg.addRecipient(Message.RecipientType.TO, toAddress);
        
        mesg.setSubject("Subject");
        
        mesg.setText("PharmacyHQ Discussion Forum "
                + " -"+ "please click to confirm email: "+ 
                "http://localhost:3000/ChangePW/" + id3 + "/" + user1 );
        
        Transport.send(mesg);
        
    System.out.println("email sent");

} else if (m9.find()) {

    int m9i = m9.group(0).indexOf("=");
    int m9i2 = m9.group(0).lastIndexOf("=");
    int m9i3 = m9.group(0).indexOf("-");
    int m9i4 = m9.group(0).lastIndexOf("-");
    
    String PW2 = m9.group(0).substring(m9i+1, m9i2);
    String email2 = m9.group(0).substring(m9i2+8, m9i3-2);
    String ID3 = m9.group(0).substring(m9i3+8, m9i4-2);
    
    String PWE2 = PE2.encode(PW2);
    
 updatePW UPW = new updatePW();
 
 UPW.now(email2, ID3, PWE2);
    
    System.out.println("Data: " + PW2+"Data: "+email2+"Data: "+ID3);
    
} else if (m10.find()) {
    
     int m10i = m10.group(0).indexOf("=");
    int m10i2 = m10.group(0).lastIndexOf("=");
    
      String email3 = m10.group(0).substring(m10i+1, m10i2);
      
      deleteUser DU = new deleteUser();
      DU.del(email3);
      
      System.out.println("Please delete: " + email3);
    
} else if (m11.find()) {
            
  fetchCommunities FC = new fetchCommunities();
  FC.Fetch();

              os.print("HTTP/1.1 200 OK" + CRLF);
              os.print("Content-Type: text/plain; charset=utf-8" + CRLF);
             os.print("Access-Control-Allow-Origin:*" + CRLF);
             os.print("Access-Control-Allow-Headers: Content-Type" + CRLF);           
            os.print("Server: http://localhost:80/" + CRLF);
            
        os.print(CRLF);
        os.print(FC.Fetch());
     
        os.flush(); //new addition
       os.close();
        }
    
            } 
             
        }catch (IOException e) { System.err.println("THIS WAS CAUGHT: " + e);} catch (SQLException ex) {
            Logger.getLogger(ServerRequests.class.getName()).log(Level.SEVERE, null, ex);
        }   catch (AddressException ex) {
                Logger.getLogger(ServerRequests.class.getName()).log(Level.SEVERE, null, ex);
            }catch (MessagingException ex) {
                Logger.getLogger(ServerRequests.class.getName()).log(Level.SEVERE, null, ex);
            }
  }).start();
        
    } 
    
}`

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by jgschmitz** [comment]
@evanr1234  here are some suggestions 

Switch to Non-Blocking I/O or Asynchronous Frameworks:

Use Java's NIO (via Selector and Channel) or frameworks like Netty, Vert.x, or Spring WebFlux for better scalability and resource efficiency.
Use a Thread Pool:

Instead of spawning a thread per connection, use a thread pool (ExecutorService) to manage threads efficiently.
Example:
```
ExecutorService threadPool = Executors.newFixedThreadPool(100);

while (true) {
    Socket clientSocket = serverSocket.accept();
    threadPool.submit(() -> handleClient(clientSocket));
}
```
Leverage a Web Framework:

Instead of writing the server from scratch, use frameworks like:
Spring Boot: For enterprise-grade APIs with tools like Spring MVC or Spring WebFlux.
Micronaut or Quarkus: For lightweight, high-performance Java applications.
Deploy on AWS with Scaling and Load Balancing:

Use an Elastic Load Balancer (ELB) to distribute traffic across multiple instances of your server.
Deploy your server on an EC2 Auto Scaling Group to handle traffic spikes by scaling instances automatically.
For even higher performance, consider serverless options like AWS Lambda for handling specific requests.
Use a Cache:

If your server frequently fetches data (e.g., from a database), use a caching layer like Redis or Memcached to reduce database load and improve response times.
